### PR TITLE
Feat/abstract configs

### PR DESCRIPTION
**Pull request by matt-winkler**

### CODE REVIEW

File: dbtc/client/cloud/configs/dbt_cloud_api.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        
+        Args:
+          data (dict): payload to create the initial request. Typically, this will be the result of a GET on the 
+             job definition from an existing job to be used for dbt Cloud migrations
+        """
+        # copy everything EXCEPT for the existing dbt Cloud job ID
+        result = self._create_job_request()
+        if data != {}:
+            for key in result.keys():
+                if key != 'id':
```
Comment (dpguthrie) [inline dbtc/client/cloud/configs/dbt_cloud_api.py:38]: @matt-winkler I was thinking about this the other day - have you ever tried creating a job (or any other resource) with an ID?  I can't imagine that our backend would actually allow this.

Comment (matt-winkler) [reply dbtc/client/cloud/configs/dbt_cloud_api.py:38]: Yeah @dpguthrie you can't include the id in the create request.

**Comment by matt-winkler** [comment]
@dpguthrie There's a lot in this PR, so to help organize thoughts I wanted to lay out a few areas where feedback would be valuable:  
* The notion of templating our API requests in [dbt_cloud_api](https://github.com/dpguthrie/dbtc/pull/38/files#diff-a378427e2a7ee9421ef0cd91aa3aea1ec053cb86f00b31f88628b7727465d797)

**Comment by matt-winkler** [comment]
@dpguthrie There's a lot in this PR, so to help organize thoughts I wanted to lay out a few areas where feedback would be valuable:

* The notion of templating our API requests in [dbt_cloud_api](https://github.com/dpguthrie/dbtc/pull/38/files#diff-a378427e2a7ee9421ef0cd91aa3aea1ec053cb86f00b31f88628b7727465d797)
* Adding the concept of `mode` to `trigger_job`. My thinking here is to make it more clear for the end user that they only get to run a job in a single mode at a time. Otherwise I could see someone passing e.g. `restart_from_failure=True` AND `autoscale=True` and be confused about what behavior to expect.
* As the number of use cases grows, the `trigger_job` method is going to have lots of possible parameters. Are we OK with that our would it be valuable to expose the concept of run modes in separate methods callable from the CLI, e.g. `dbtc trigger-job-with-restart-from-failure`, `dbtc trigger-job-with-autoscale`, etc.

**Comment by dpguthrie** [comment]
> The notion of templating our API requests in [dbt_cloud_api](https://github.com/dpguthrie/dbtc/pull/38/files#diff-a378427e2a7ee9421ef0cd91aa3aea1ec053cb86f00b31f88628b7727465d797)

I came up with a different approach for validation using the [pydantic]() library.  This would both do validation of the payload that a user specifies as well as give it the appropriate defaults (`id` for instance).

https://github.com/dpguthrie/dbtc/pull/39

We can iterate on what's best, I just didn't want to overwrite all your work.

**Comment by dpguthrie** [comment]
> As the number of use cases grows, the trigger_job method is going to have lots of possible parameters. Are we OK with that or would it be valuable to expose the concept of run modes in separate methods callable from the CLI, e.g. dbtc trigger-job-with-restart-from-failure, dbtc trigger-job-with-autoscale, etc.

Putting this here as well - my take is that we should be building out different methods that accomplish these different tasks.  The `trigger_job` method is already probably a bit too big and doing too much.  It should be the entry point to trigger a job from all those other methods, but each of those methods will have logic that doesn't make sense to wrap all together.  Also, I'm guessing you're just gonna continue thinking of awesome use cases for this library, so we should plan for that.

**Comment by matt-winkler** [comment]
@dpguthrie 100 on breaking out separate methods that all get to the base trigger_job. I was mid-way on that before thinking to stop and ask you :))

File: dbtc/client/cloud/base.py [hunk 2]
```diff
@@ -77,10 +59,28 @@ def _header_property(self):
 
         return 'api_key'
 
+    def _clone_resource(self, resource: str, **kwargs):
```
Comment (matt-winkler) [inline dbtc/client/cloud/base.py:62]: @dpguthrie this is very cool! Is there a means I could use to subset which fields from the base object are used for replication? It seems like this is the intent of `create_args`, but I think a field subset being passed there would need to affect the `payload` in order for that modification to occur.

Context: When trying to replicate a job, I'm not able to use fields such as `deferring_job_definition_id` on the CREATE JOB endpoint. The process needs to first create the job with fewer fields, and then add the `deferring_job_defintion_id` in an update.

Comment (dpguthrie) [reply dbtc/client/cloud/base.py]: Ya, we need another argument here for that.  Good call!  Who knows this private method may be a bit overkill for the things we need to do as well as thinking about what resources actually need to be cloned.

Comment (matt-winkler) [reply dbtc/client/cloud/base.py]: Per conversation offline, noting that we agreed this isn't something to pursue further for now. There will be too much situational variation in what's needed to replicate resources, so we should stick to the CRUD operations and adjust as-needed for other use cases.

File: dbtc/client/cloud/base.py [hunk 3]
```diff
@@ -1002,6 +1035,279 @@ def test_connection(self, account_id: int, payload: Dict) -> Dict:
...
+            # fill in optional fields on the job if present in the source
+            for field in optional_fields:
+                new_job_definition[field] = existing_job_defintion.get(field)
+            
+            # set the ID in the update request
+            new_job_definition['id'] = job_id
+            self.console.log(new_job_definition)
+            job_id = self.update_job(
+                account_id=account_id, job_id=job_id, payload=new_job_definition
+            )['data']['id']
```
Comment (dpguthrie) [inline dbtc/client/cloud/base.py:1240]: @matt-winkler I'm not sure I follow here.  Why wouldn't we just create the new job from the existing job definition (outside of setting `id` to `None` and changing the `name`)?

Comment (matt-winkler) [reply dbtc/client/cloud/base.py:1240]: @dpguthrie we can't copy the existing data in the job 1:1 in a new CREATE JOB request. For example, you can't create a new job including the `deferring_job_definition_id` parameter.

Comment (matt-winkler) [reply dbtc/client/cloud/base.py:1240]: @dpguthrie updated my comment with a better example

**Review by matt-winkler — COMMENTED** [review]
🔥

**Comment by matt-winkler** [comment]
@dpguthrie Think I found a happier medium with the cloning idea. This updates _clone_resource so that it will return a payload that includes the object's required fields + and optional fields that are not `None` in the source data. Then we can further modify depending on the use case e.g. 1) in autoscaling we want to add an identifier to the job's name OR 2) if this was for migration, just push the same payload to another account / project etc.

**Comment by matt-winkler** [comment]
Also adding a comment that tests on autoscaling jobs are failing because we need to standardize on an account / jobs to work from.

**Pull request closed without merging**